src/libpriv/rpmostree-rpm-util.c

```diff
@@ -1257,6 +1257,57 @@ rpmostree_create_rpmdb_pkglist_variant (int              dfd,
   return TRUE;
 }
 
+gboolean
+rpmostree_create_pkglist_variant (GPtrArray     *pkglist,
+                                  GVariant     **out_variant,
+                                  GCancellable  *cancellable,
+                                  GError       **error)
+{
+  /* we insert it sorted here so it can efficiently be searched on retrieval */
+  g_ptr_array_sort (pkglist, (GCompareFunc)rpmostree_pkg_array_compare);
+
+  GVariantBuilder pkglist_v_builder;
+  g_variant_builder_init (&pkglist_v_builder, G_VARIANT_TYPE ("a(ss)"));
+  for (guint i = 0; i < pkglist->len; i++)
+    {
+      DnfPackage *pkg = pkglist->pdata[i];
+      g_autofree gchar *repodata_chksum = NULL;
+
+      if (!rpmostree_get_repodata_chksum_repr (pkg, &repodata_chksum, error))
+        return FALSE;
+
+      g_variant_builder_add (&pkglist_v_builder, "(ss)",
+                             dnf_package_get_nevra (pkg),
+                             repodata_chksum);
+    }
+
+  *out_variant = g_variant_ref_sink (g_variant_builder_end (&pkglist_v_builder));
+  return TRUE;
+}
+
+DnfPackage *
+rpmostree_get_locked_package (DnfSack    *sack,
+                              GHashTable *lockmap,
+                              const char *name)
+{
+  /* The manifest might specify not only package names (foo-1.x) but also
+   * something-that-foo-provides */
+  g_autoptr(GPtrArray) alts = rpmostree_get_matching_packages (sack, name);
```

Comment:
Hmm, I know I suggested this, but I'm not completely satisfied either with how "indirect" it all seems. One root issue here of course is #731; libsolv simply doesn't make it easy to provide resolution information. (It's not entirely its fault either; depsolving is a hard problem that isn't exactly disposed to provide clear cut answers to these sorts of questions).

Anyway, good to roll with this for now! Though it might need more thinking and tweaks in the future.

Reply:
Just couldn't help thinking about this a bit more :)

ISTM like in the lockfile case, maybe we do want to require manifests to not use "provides" but instead purely pkgnames. So e.g. when using `--lockfile` or `--write-lockfile-to`, we would only search by pkgname rather than the default `dnf_context_install()` fuzzy logic and error out if the pkgname isn't found. That would really simplify the semantics of the lockfile.

But yeah, I'm happy getting this in as is for now, and then tweak it in follow-ups!

Reply:
The tiny downside of course is that it requires slightly more churn on treefile writers to react to things like `Obsoletes`, package splits, etc... Doesn't seem like an unreasonable burden though.

Reply:
Follow-up to this in #1849.